### PR TITLE
Implement Password Reset Feature

### LINKED ISSUE #58: Implement password reset endpoints
**Description:**
In order to enhance the security and user-friendliness of our authentication system, we are introducing a crucial feature: the Password Reset functionality. 

**Tasks:**
1. Create the following routes here `src/routes/auth/auth.route.ts`

`/password-reset-request`
`/password-reset`

2. Create the services here `src/services/auth.service.ts`
3. Create the controllers here `src/services/auth.controller.ts`
4. `/password-reset-request` will check the user email and generate a reset token. for now, return the token. In the future password reset link will be emailed to the user. The reset token should include the user id and an expiration time of 1h.
5. `/password-reset` will verify the token (check the user id and the expiration time) and update the new password

**Related Dependencies or References:**
- Passport.js: [link to documentation](https://www.passportjs.org/)
- JWT: [link to documentation](https://jwt.io/)

### PR DESCRIPTION
**Pull request by mayura-andrew**
## Purpose

The purpose of this PR is to fix #58 

## Goals
- Fix the bug in the password reset feature
- Improve the user experience when resetting passwords

## Approach
Used JWT to implement this password reset password

### Screenshots
N/A

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

## Test environment
- OS : Fedora Linux 39 
- Node version : v20.10.0

## Learning
Learned how to implement password reset API endpoint in backend

### CODE REVIEW

File: src/services/auth.service.ts [hunk 1]
```diff
@@ -108,3 +108,64 @@ export const findOrCreateUser = async (
...
+  await dataSource.getRepository(Profile).save(profile)
+}
+
+export const resetPassword = async (
+  token: string,
+  newPassword: string
+): Promise<{ statusCode: number; message: string }> => {
+  try {
+    const decoded = jwt.verify(token, JWT_SECRET) as { userId: string }
+    console.log('decoded', decoded)
```
Comment (anjula-sack) [inline src/services/auth.service.ts:153]:
```suggestion
```
Can you write the tests as well?

**Review by anjula-sack — CHANGES_REQUESTED** [review]

**Comment by mayura-andrew** [comment]
> Looks good to me @mayura-andrew Can you write tests for this?

**Comment by mayura-andrew** [comment]
@anjula-sack hey can you check test cases?

File: package.json [hunk 2]
```diff
@@ -22,6 +22,7 @@
     "dotenv": "^16.3.1",
     "ejs": "^3.1.10",
     "express": "^4.18.2",
+    "express-session": "^1.18.0",
```
Comment (anjula-sack) [inline package.json:25]: Why this is intalled?

File: src/services/auth.service.test.ts [hunk 3]
```diff
@@ -159,3 +164,54 @@ describe('loginUser', () => {
...
+
+describe('Auth Service', () => {
+  let token: any
+  const validEmail = 'test@gmail.com'
+  const invalidEmail = 'invalid@ousl.lk'
+  const newPassword = 'newpassword123'
+
+  beforeAll(async () => {
+    token = await generateResetToken(validEmail)
+    console.log(token)
```
Comment (anjula-sack) [inline src/services/auth.service.test.ts:176]:
```suggestion
```

File: src/services/auth.service.ts [hunk 4]
```diff
@@ -101,3 +108,104 @@ export const findOrCreateUser = async (
...
+      return {
+        statusCode: 401,
+        message: 'Invalid email or password',
+        token: ''
+      }
+    }
+    const token = jwt.sign({ userId: profile.uuid }, JWT_SECRET, {
+      expiresIn: '1h'
+    })
+    // Getting the email content
```
Comment (anjula-sack) [inline src/services/auth.service.ts:132]: Remove comments
```suggestion
```

File: src/services/auth.service.ts [hunk 5]
```diff
@@ -101,3 +108,104 @@ export const findOrCreateUser = async (
...
+    const content = getPasswordResetEmailContent(email, token)
+    // Sending the email with token
```
Comment (anjula-sack) [inline src/services/auth.service.ts:134]:
```suggestion
```

File: src/services/auth.service.ts [hunk 6]
```diff
@@ -101,3 +108,104 @@ export const findOrCreateUser = async (
...
+      statusCode: 200,
+      message: 'Password reset link successfully sent to email',
+      token
+    }
+  } catch (error) {
+    console.error(
+      'Error executing Reset Password && Error sending password reset link',
+      error
+    )
+    return { statusCode: 500, message: 'Internal server error', token: '' }
```
Comment (anjula-sack) [inline src/services/auth.service.ts:148]:
```suggestion
    return { statusCode: 500, message: 'Internal server error'}
```

File: src/services/auth.service.ts [hunk 7]
```diff
@@ -101,3 +108,104 @@ export const findOrCreateUser = async (
...
+  }
+
+  const profileRepository = dataSource.getRepository(Profile)
+  const profile = await profileRepository.findOne({
+    where: { uuid: decoded.userId }
+  })
+
+  if (!profile) {
+    console.error('Error executing Reset Password: No profile found')
+    return { statusCode: 404, message: 'No profile found' }
```
Comment (anjula-sack) [inline src/services/auth.service.ts:188]:
```suggestion
    return { statusCode: 409, message: 'No profile found' }
```

File: src/services/auth.service.ts [hunk 8]
```diff
@@ -101,3 +108,104 @@ export const findOrCreateUser = async (
...
+  if (!token || !newPassword) {
+    console.error('Error executing Reset Password: Missing parameters')
+    return { statusCode: 400, message: 'Missing parameters' }
+  }
+
+  let decoded
+  try {
+    decoded = jwt.verify(token, JWT_SECRET) as { userId: string }
+  } catch (error) {
+    console.error('Error executing Reset Password: Invalid token', error)
```
Comment (anjula-sack) [inline src/services/auth.service.ts:177]: Just throw the error. It will be catched by the try catch

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (mayura-andrew) [reply package.json:25]: oops i forgot remove it sorry

File: src/services/auth.service.ts (hunk 8 above)
Comment (mayura-andrew) [reply src/services/auth.service.ts:177]: like this   ```
let decoded
  try {
    decoded = jwt.verify(token, JWT_SECRET) as { userId: string }
  } catch (error) {
    throw new Error('Invalid token')
  }
```

File: src/services/auth.service.test.ts [hunk 9]
```diff
@@ -166,8 +166,8 @@ describe('loginUser', () => {
 })
 
 describe('Auth Service', () => {
-  let token: any
-  const validEmail = 'test@gmail.com'
+  let token: string | any
```
[the remark this answers is not in the document]
Comment (mayura-andrew) [reply src/services/auth.service.test.ts]: i update it

File: package.json [hunk 10]
```diff
@@ -40,6 +41,7 @@
     "@types/cors": "^2.8.13",
     "@types/ejs": "^3.1.5",
     "@types/express": "^4.17.17",
+    "@types/express-session": "^1.18.0",
```
Comment (anjula-sack) [inline package.json:44]:
```suggestion
```
Can you remove this as well?

File: src/controllers/auth.controller.ts [hunk 11]
```diff
@@ -157,3 +162,47 @@ export const requireAuth = (
...
+    const { token, newPassword } = req.body
+
+    if (!token || !newPassword) {
+      res
+        .status(400)
+        .json({ error: 'Token and new password are required fields' })
+      return
+    }
+    const result = await resetPassword(token, newPassword)
+    res.status(result.statusCode).json({ message: result.message })
```
Comment (anjula-sack) [inline src/controllers/auth.controller.ts:200]:
```suggestion
    const { statusCode, message } = await resetPassword(token, newPassword)
    res.status(statusCode).json({ message })
```

File: src/routes/auth/auth.route.ts [hunk 12]
```diff
@@ -21,4 +23,7 @@ authRouter.get(
 )
 
 authRouter.get('/google/callback', googleRedirect)
+
```
Comment (anjula-sack) [inline src/routes/auth/auth.route.ts:26]:
```suggestion
```

File: src/types.ts [hunk 13]
```diff
@@ -26,3 +26,14 @@ export interface ApiResponse<T> {
 export interface User extends Express.User {
   primary_email: string
 }
+
+export interface ResetToken {
+  statusCode: number
+  message: string
+  token?: string
+}
```
Comment (anjula-sack) [inline src/types.ts:34]: Use the existing APIResponse interface

File: src/types.ts [hunk 14]
```diff
@@ -26,3 +26,14 @@ export interface ApiResponse<T> {
...
+
+export interface ResetPassword {
+  statusCode: number
+  message: string
+}
```
Comment (anjula-sack) [inline src/types.ts:39]: Same here.

File: src/services/auth.service.ts [hunk 15]
```diff
@@ -101,3 +109,100 @@ export const findOrCreateUser = async (
...
+        content.message
+      )
+    }
+  } catch (error) {
+    console.error(
+      'Error executing Reset Password: Error updating profile',
+      error
+    )
+    return { statusCode: 500, message: 'Error updating profile' }
+  }
```
Comment (anjula-sack) [inline src/services/auth.service.ts:205]: Remove this try catch.

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: package.json (hunk 10 above)
Comment (mayura-andrew) [reply package.json:44]: okay sorry i forgot to remove that

**Review by anjula-sack — APPROVED** [review]
Good job! @mayura-andrew

**Pull request merged**